lib/src/command/upgrade.dart

```diff
@@ -35,23 +47,241 @@ class UpgradeCommand extends PubCommand {
+  bool get _shouldShowSpinner => stdout.hasTerminal;
+
+  bool get _dryRun => argResults['dry-run'];
+
   @override
   Future<void> runProtected() async {
     if (argResults.wasParsed('packages-dir')) {
       log.warning(log.yellow(
           'The --packages-dir flag is no longer used and does nothing.'));
     }
+
+    if (argResults['nullsafety'] || argResults['null-safety']) {
+      return await _runUpgradeNullSafety();
+    }
+
     await entrypoint.acquireDependencies(SolveType.UPGRADE,
         useLatest: argResults.rest,
-        dryRun: argResults['dry-run'],
+        dryRun: _dryRun,
         precompile: argResults['precompile']);
 
+    _showOfflineWarning();
+  }
+
+  Future<void> _runUpgradeNullSafety() async {
+    final upgradeOnly = [
+      if (argResults.rest.isNotEmpty)
+        ...argResults.rest
+      else ...[
+        ...entrypoint.root.pubspec.dependencies.keys,
+        ...entrypoint.root.pubspec.devDependencies.keys
+      ],
+    ];
+
+    final nullsafetyPubspec = await _upgradeToNullSafetyConstraints(
+      entrypoint.root.pubspec,
+      upgradeOnly,
+    );
+
+    /// Solve [nullsafetyPubspec] in-memory and consolidate the resolved
+    /// versions of the packages into a map for quick searching.
+    final resolvedPackages = <String, PackageId>{};
+    await log.spinner('Resolving dependencies', () async {
+      final solveResult = await tryResolveVersions(
+        SolveType.UPGRADE,
+        cache,
+        Package.inMemory(nullsafetyPubspec),
+      );
+      for (final resolvedPackage in solveResult?.packages ?? []) {
+        resolvedPackages[resolvedPackage.name] = resolvedPackage;
+      }
+    }, condition: _shouldShowSpinner);
+
+    /// List of changes to made to `pubspec.yaml`.
+    final changes = <PackageRange, PackageRange>{};
+    final declaredHostedDependencies = [
+      ...entrypoint.root.pubspec.dependencies.values,
+      ...entrypoint.root.pubspec.devDependencies.values,
+    ].where((dep) => dep.source is HostedSource);
+    for (final dep in declaredHostedDependencies) {
+      final resolvedPackage = resolvedPackages[dep.name];
+      assert(resolvedPackage != null);
+      if (resolvedPackage == null || !upgradeOnly.contains(dep.name)) {
+        // If we're not to upgrade this package, or it wasn't in the
+        // resolution somehow, then we ignore it.
+        continue;
+      }
+
+      final constraint = VersionConstraint.compatibleWith(
+        resolvedPackage.version,
+      );
+      if (dep.constraint.allowsAll(constraint) &&
+          constraint.allowsAll(dep.constraint)) {
+        // If constraint allows the same as the existing constraint then
+        // there is no need to changes.
+        continue;
+      }
+
+      changes[dep] = dep.withConstraint(constraint);
+    }
+
+    if (!_dryRun) {
+      await _updatePubspec(changes);
+
+      // TODO: Allow Entrypoint to be created with in-memory pubspec, so that
```

Comment:
We can maybe do a call to  `Future<SolveResult> tryResolveVersions(` to get the list of updates, and then call `result.summarizeChanges(type, dryRun: dryRun);`

Reply:
I tried that... I couldn't get it to do what I wanted.

I think it was because it didn't read the lock file, for now I think we can just keep it simple.
In fact we could argue that command that change the `pubspec.yaml` should focus the dry-run output on what they would have changed in `pubspec.yaml`. All the other changes in `pubspec.lock` are rather uninteresting, and distract from the important bits (what is changed in the `pubspec.yaml`).